### ISSUE TITLE
Cross-checking with blockchain data

### ISSUE THREAD

**Issue by cassiniNMC** [namecoin/namecoin-canaries#3]
Due to my pretty limited pgp skills I can think of only this method of cross-checking the signature verification with uncensorable information on the Namecoin blockchain:

```
./namecoin-cli name_show id/domob
---> d•domob.eu, fpr:1142850E6DFF65BA63D688A8B2492AC4A7330737
curl 'https://sks-keyservers.net/pks/lookup?op=get&options=mr&search=0x1142850E6DFF65BA63D688A8B2492AC4A7330737' | gpg --import
---> ... "Daniel Kraft <d•domob.eu>" ...
gpg --verify  canary-001-2017.md.sig.domob  canary-001-2017.md
---> gpg: Good signature from "Daniel Kraft <d•domob.eu>"
---> Primary key fingerprint: 1142 850E 6DFF 65BA 63D6  88A8 B249 2AC4 A733 0737
```

```
./namecoin-cli name_show id/hlandau
---> hlandau•devever.net, fpr:95A9908DDFA16830BE9FB9003D30A3A9FF1360DC
curl 'https://sks-keyservers.net/pks/lookup?op=get&options=mr&search=0x95A9908DDFA16830BE9FB9003D30A3A9FF1360DC' | gpg --import
---> "Hugo Landau (2013 Principal) <hlandau•devever.net>"
gpg --verify  canary-001-2017.md.sig.hlandau  canary-001-2017.md
---> Good signature from "Hugo Landau (2013 Principal) <hlandau•devever.net>"
---> Primary key fingerprint: 95A9 908D DFA1 6830 BE9F  B900 3D30 A3A9 FF13 60DC
```

Both Daniel's and Hugo's data look consistent. When I try verifying @JeremyRand's signature I run into issues:

```
./namecoin-cli name_show id/jeremy
---> jeremy•veclabs.net, fpr:9B0FCA28C8A773B514B4775801CDFC46D9D72D3A
curl 'https://sks-keyservers.net/pks/lookup?op=get&options=mr&search=0x9B0FCA28C8A773B514B4775801CDFC46D9D72D3A' | gpg --import
---> "Jeremy Rand <jeremy•veclabs.net>"
gpg --verify  canary-001-2017.md.sig.jeremy  canary-001-2017.md
---> "jeremyrand•airmail.cc"
---> Can't check signature: No public key
```

How can I tell (or anybody else trying to monitor the canary's health) that jeremy•veclabs.net and jeremyrand•airmail.cc are identical?

**Comment by JeremyRand** [comment]
> How can I tell (or anybody else trying to monitor the canary's health) that jeremy•veclabs.net and jeremyrand•airmail.cc are identical?

Good point.  I switched OpenPGP keys a while back (I think last year?) due to getting a more secure setup (the new private key is generated inside an isolated offline Qubes VM that has never had any data imported into it).  I suspect that I may have neglected to update the `id/jeremy` name when I did this key switch.  I'll look into fixing the data in `id/jeremy`.  That said... does the new key have any signatures from the old key?  I *think* I signed the new key with the old key... if I failed to do that, let me know and I'll see if I can fix that.

**Comment by JeremyRand** [comment]
By the way, don't use `curl` for downloading the keys; doing so trusts the keyserver to give you the key you actually wanted.  GPG has a `recv-keys` command (at least, I think that's the syntax), which actually verifies that the fingerprint of the downloaded key matches the requested fingerprint.

**Comment by JeremyRand** [comment]
> I'll look into fixing the data in id/jeremy.

Should be fixed as soon as the transaction gets mined.

**Comment by cassiniNMC** [comment]
> By the way, don't use curl for downloading the keys; doing so trusts the keyserver to give you the key you actually wanted. GPG has a recv-keys command (at least, I think that's the syntax), which actually verifies that the fingerprint of the downloaded key matches the requested fingerprint.

I've just come across a flashy red warning label at https://wiki.archlinux.org/index.php/GnuPG next to the --recv-keys usage instructions:

>> To import a key from a key server: gpg --recv-keys key-id
>> **Warning: You should verify the authenticity of the retrieved public key by comparing its fingerprint ...**

This indicates that --recv-keys does not verify the fingerprints. Could easily be there is some exotic extra parameter that forces gpg doing so. Either way, both curl or gpg --recv-keys require a subsequent visual inspection using gpg --fingerprint, I guess.

There is something much more disturbing about --recv-keys, though.
_tcpdump_ just told me that gpg --recv-keys communicates in plain text in both directions:

`... GET /pks/lookup?op=get&options=mr&search=0xB3F2D165786D6570 HTTP/1.0 ...`

Again, there might be a nifty parameter that encourages gpg in applying at least a small degree of obfuscation here.

Therefore I feel more comfortable using curl https://... instead of --recv-keys. Even more so as curl -v gives me some extra pieces of information on what happens under the bonnet.

**Comment by cassiniNMC** [comment]
> I'll look into fixing the data in id/jeremy.
> Should be fixed as soon as the transaction gets mined.

ok, looks much better now:

```
./namecoin-cli name_show id/jeremy
---> jeremyrand•airmail.cc, fpr:51740B7C732D572A31404010660555E1F8F7BF85
curl -v 'https://sks-keyservers.net/pks/lookup?op=get&options=mr&search=0x51740B7C732D572A31404010660555E1F8F7BF85' | gpg --import
gpg --fingerprint
---> ...
---> 5174 0B7C 732D 572A 3140  4010 6605 55E1 F8F7 BF85
---> Jeremy Rand (Master Key) <jeremy•veclabs.net>
---> Jeremy Rand <jeremyrand•airmail.cc>
---> ...
gpg --verify canary-001-2017.md.sig.jeremy  canary-001-2017.md
---> Good signature from "Jeremy Rand (Master Key) <jeremy•veclabs.net>"
                     aka "Jeremy Rand <jeremyrand•airmail.cc>"  ...
---> Primary key fingerprint: 5174 0B7C 732D 572A 3140  4010 6605 55E1 F8F7 BF85
```

&nbsp;
> I think I signed the new key with the old key... if I failed to do that, let me know
>
I have no idea how to sign a new key with an old one, let alone how to check if someone else did so. It doesn't matter anyway, I think. The updated id/jeremy blockchain data is consistent with the results of gpg --verify, and there is no error message anymore 😅 💦
